tests/linearizability/history.go

```diff
@@ -0,0 +1,119 @@
+	id         int
+	idProvider idProvider
+
+	history
+}
+
+func newAppendableHistory(ids idProvider) *appendableHistory {
+	return &appendableHistory{
+		id:         ids.ClientId(),
+		idProvider: ids,
+		history: history{
+			successful: []porcupine.Operation{},
+			failed:     []porcupine.Operation{},
+		},
+	}
+}
+
+func (h *appendableHistory) AppendGet(key string, start, end time.Time, resp *clientv3.GetResponse) {
+	var readData string
+	if len(resp.Kvs) == 1 {
+		readData = string(resp.Kvs[0].Value)
+	}
+	h.successful = append(h.successful, porcupine.Operation{
+		ClientId: h.id,
+		Input:    EtcdRequest{Op: Get, Key: key},
+		Call:     start.UnixNano(),
+		Output:   EtcdResponse{GetData: readData, Revision: resp.Header.Revision},
+		Return:   end.UnixNano(),
+	})
+}
+
+func (h *appendableHistory) AppendPut(key, value string, start, end time.Time, resp *clientv3.PutResponse, err error) {
+	if err != nil {
+		h.failed = append(h.failed, porcupine.Operation{
+			ClientId: h.id,
+			Input:    EtcdRequest{Op: Put, Key: key, PutData: value},
+			Call:     start.UnixNano(),
+			Output:   EtcdResponse{Err: err},
+			Return:   0, // For failed writes we don't know when request has really finished.
+		})
+		// Operations of single client needs to be sequential.
+		// As we don't know return time of failed operations, all new writes need to be done with new client id.
+		h.id = h.idProvider.ClientId()
+		return
+	}
+	var revision int64
+	if resp != nil && resp.Header != nil {
+		revision = resp.Header.Revision
+	}
+	h.successful = append(h.successful, porcupine.Operation{
+		ClientId: h.id,
+		Input:    EtcdRequest{Op: Put, Key: key, PutData: value},
+		Call:     start.UnixNano(),
+		Output:   EtcdResponse{Err: err, Revision: revision},
+		Return:   end.UnixNano(),
+	})
+}
+
+type history struct {
+	successful []porcupine.Operation
+	// failed requests are kept separate as we don't know return time of failed operations.
+	failed []porcupine.Operation
+}
+
+func (h history) Merge(h2 history) history {
+	result := history{
+		successful: make([]porcupine.Operation, 0, len(h.successful)+len(h2.successful)),
+		failed:     make([]porcupine.Operation, 0, len(h.failed)+len(h2.failed)),
+	}
+	result.successful = append(result.successful, h.successful...)
+	result.successful = append(result.successful, h2.successful...)
+	result.failed = append(result.failed, h.failed...)
+	result.failed = append(result.failed, h2.failed...)
+	return result
+}
+
+func (h history) Operations() []porcupine.Operation {
+	operations := make([]porcupine.Operation, 0, len(h.successful)+len(h.failed))
+	var maxTime int64
+	for _, op := range h.successful {
+		operations = append(operations, op)
+		if op.Return > maxTime {
+			maxTime = op.Return
+		}
+	}
+	for _, op := range h.failed {
+		if op.Call > maxTime {
+			continue
+		}
+		op.Return = maxTime + 1
+		operations = append(operations, op)
+	}
```

Comment:
If I understand correctly, you followed the suggestion in https://github.com/anishathalye/porcupine/issues/10? If yes, please add a comment here, otherwise it's hard for other contributors to understand.

Reply:
Added the comment